src/modules/roc_pipeline/target_stdio/roc_pipeline/parse_port.cpp

```diff
@@ -47,97 +47,140 @@ bool match_proto(PortType type, const char* str, PortProtocol& proto) {
     return false;
 }
 
-} // namespace
+PortProtocol parse_proto(const char* begin, const char* end, PortType type) {
+    char proto_buf[16] = {};
 
-bool parse_port(PortType type, const char* input, PortConfig& result) {
-    if (input == NULL) {
-        roc_log(LogError, "parse port: string is null");
-        return false;
+    if (size_t(end - begin) > sizeof(proto_buf) - 1) {
+        roc_log(LogError, "parse port: bad protocol: too long");
+        return Proto_None;
     }
 
-    const char* lcolon = strchr(input, ':');
-    const char* rcolon = strrchr(input, ':');
+    memcpy(proto_buf, begin, size_t(end - begin));
+    proto_buf[end - begin] = '\0';
 
-    if (!lcolon || !rcolon || lcolon == rcolon || lcolon == input || !rcolon[1]) {
-        roc_log(LogError,
-                "parse port: bad format: expected PROTO:ADDR:PORT or PROTO::PORT");
+    PortProtocol protocol = Proto_None;
+    if (!match_proto(type, proto_buf, protocol)) {
+        return Proto_None;
+    }
+
+    return protocol;
+}
+
+long parse_port_num(const char* port) {
+    if (!isdigit(*port)) {
+        roc_log(LogError, "parse port: bad port: not a number");
+        return -1;
+    }
+
+    char* port_end = NULL;
+    const long port_num = strtol(port, &port_end, 10);
+
+    if (port_num == LONG_MAX || port_num == LONG_MIN || !port_end || *port_end) {
+        roc_log(LogError, "parse port: bad port: not a positive integer");
+        return -1;
+    }
+
+    if (port_num < 0 || port_num > 65535) {
+        roc_log(LogError, "parse port: bad port: not in range [1; 65535]");
+        return -1;
+    }
+
+    return port_num;
+}
+
+bool parse_ipv4_addr(const char* addr, long port_num, PortConfig& result) {
+    if (!result.address.set_ipv4(addr, (int)port_num)) {
+        roc_log(LogError, "parse port: bad IPv4 address: %s", addr);
         return false;
     }
 
-    PortProtocol protocol = Proto_None;
+    return true;
+}
 
-    char proto_buf[16] = {};
-    if (size_t(lcolon - input) > sizeof(proto_buf) - 1) {
-        roc_log(LogError, "parse port: bad protocol: too long");
+bool parse_ipv6_addr(const char* addr, long port_num, PortConfig& result) {
+    const size_t addrlen = strlen(addr);
+
+    if (addr[addrlen - 1] != ']') {
```

Comment:
It would be better to work with `[` and `]` in the same function instead of spreading it across several. I suggest to work with brackets in parse_addr() and pass begin/end pointers to parse_ipv6_addr().

Reply:
Actually, to unify all these helpers, I would suggest to pass begin/end pointers to *every* helper. Currently some of them accept begin/end pointers, and some accept null-terminated strings, which is a bit confusing.

Reply:
Fixed